### Regression(회귀) 실습: [**나이, 성별, BMI, 자녀, 흡연여부**]에 따른 보험료(insurance) 산정값 예측

#### 4 종류 알고리즘의 구현 및 비교

1) Linear

2) Decision tree

3) Random forest

4) XGBoost


In [1]:
from IPython.display import Image
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings

# 경고 메시지 출력 표기 생략
warnings.filterwarnings('ignore')

In [ ]:
# 보험 데이터셋 읽어오기
data = pd.read_csv("https://raw.githubusercontent.com/agtechresearch/LectureAlgorithm/main/csv/insurance.csv")
data

# gender 는 남성/여성 -> 1은 남성, 0은 여성
# smoker 는 흡연/비흡연 -> 1은 흡연, 0은 비흡연

In [ ]:
data.head()

In [ ]:
# 데이터셋 정보 출력


In [ ]:
# 통계 정보 출력


In [ ]:
# 통계 정보 소수점 2자릿수까지 반올림


### 데이터셋, 독립변수와 종속변수 분리: 독립변수 -> x, 종속변수 -> y

In [ ]:
x =
y =

In [ ]:
# 독립변수 x 확인


In [ ]:
# 종속변수 y 확인


### 학습용 데이터셋(train set)과 테스트셋(test set)으로 데이터 분할<br>(몇몇 논문과 도서 등에서는 검증셋 validation set 이라고도 함)

학습용 데이터셋(train set) ==> X_train(독립변수), y_train(종속변수)

테스트용/검증용 데이터셋(test set) ==> X_test(독립변수), y_test(종속변수)

In [4]:
Image(url='https://raw.githubusercontent.com/agtechresearch/LectureAlgorithm/main/img/W05_datasplit.png', width=700)

In [5]:
Image(url='https://raw.githubusercontent.com/agtechresearch/LectureAlgorithm/main/img/W05_train_test.png', width=600)

In [ ]:
from sklearn.model_selection import train_test_split

# 데이터셋 분할 하기
X_train, X_test, y_train, y_test =

---------------

### 알고리즘: 1) Linear

#### <선형회귀 장단점>

장점
 - 모델이 간단하기 때문에 구현과 해석이 쉽다
 - 모델링하는 데 오랜 시간이 걸리지 않는다

단점
 - 최신 알고리즘에 비해 예측력이 떨어진다
 - 독립변수와 예측변수의 선형 관계를 전제로 하기 때문에, 이러한 전제에서 벗어나는 데이터에서는 좋은 예측을 보여주기 어렵다

In [ ]:
from sklearn.linear_model import LinearRegression

# 모델 불러오기
model_linear =
# 모델 훈련하기


In [ ]:
# 모델을 활용하여 예측 및 평가하기
pred_linear =

In [ ]:
# 예측 모델 평가하기
eval_linear = pd.DataFrame({'actual':______, 'pred_linear':______})
eval_linear

In [ ]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

# RMSE 값 출력 (** 0.5 는 루트를 의미 --> 0.5 를 제곱으로 계산하겠다는 의미)
print('RMSE = ', )

# 다른방법으로의 RMSE 값 출력 (squared = False 는 루트를 의미)
# mean_squared_error(y_test, pred, squared = False)

# R2 값 출력
print('R2 = ',)

In [ ]:
plt.figure(figsize=(6,6))
sns.scatterplot(x = 'actual', y = 'pred_linear', data = eval_linear)

### 이해하기 : 선형 회귀 (Linear Regression)

In [ ]:
model_linear.coef_

In [6]:
Image(url='https://raw.githubusercontent.com/agtechresearch/LectureAlgorithm/main/img/W05_insurance_fx.png', width=750)

In [ ]:
pd.Series(model_linear.coef_, index = x.columns)

age           264.799803
gender         17.344661
bmi           297.514806
children      469.339602
smoker      23469.280173
dtype: float64

<위의 내용에 대한 해석>

- age가 1만큼 증가하면 charges는 약 265만큼 증가
- gender는 0과 1로만 구성된 데이터이기 때문에 ‘1만큼 증가할 때 charges가 17만큼 증가한다기’보다는,<br>남자(1)의 경우 여자(0)보다 charges가 보통 17정도 높다’고 해석가능
- smoker도 gender와 마찬가지로 해석할 수 있습니다.

In [ ]:
model_linear.intercept_

-11576.999976112338

### 위의 정보를 이용하여 1차원 수식(linear) 완성

### charges = 264.799803 * age + 17.344661 * gender + 297.514806 * bmi + 469.339602 * children + 23469.280173 * smoker - 11576.999976112367


---------------

### 알고리즘: 2) Decision tree

#### <Decision tree 장단점>

장점
 - 데이터에 대한 가정이 없음(non-parametric model):<br>예) 선형 모델의 경우 정규분포에 대한 가정이나 독립변수-종속변수의 선형 관계를 가정으로 모델을 만들지만,<br>결정트리는 데이터에 대한 가정이 없으므로 어디에나 자유롭게 적용할 수 있음
 - 아웃라이어(outlier, 이상치)에 영향을 거의 받지 않음
 - Tree-graph 를 통해 직관적으로 이해하고 설명할 수 있음, 시각화에 탁월함


단점
 - 입력 데이터의 작은 변동에도 트리의 구성이 크게 달라질 수 있음
 - 트리가 무한정 깊어지면 오버피팅 문제를 야기할 수 있음
 - Random forest, XGBoost, LightGBM 등과 같은 트리기반의 모델들에 비해 예측력이 떨어짐

In [ ]:
from sklearn.tree import DecisionTreeRegressor

# 모델 불러오기
model_dt =
# 모델 훈련하기


In [ ]:
# 학습된 모델을 사용하여 예측 및 평가하기
pred_dt =

# RMSE, R2 값 출력 (** 0.5 는 루트를 의미 --> 0.5 를 제곱으로 계산하겠다는 의미)
print('RMSE = ', )
print('R2 = ', )


---------------

### 알고리즘: 3) Random forest

#### <Random forest 장단점>

장점
 - 결정트리와 마찬가지로 아웃라이어에 거의 영향을 받지 않음
 - 선형/비선형 데이터에 상관없이 잘 작동함


단점
 - 학습 속도가 상대적으로 느림
 - 수많은 트리를 생성하고 사용하기 때문에 모델에 대한 해석이 어려움

In [ ]:
from sklearn.ensemble import RandomForestRegressor

# 모델 불러오기
model_rf =
# 모델 훈련하기


In [ ]:
# 학습된 모델을 사용하여 예측 및 평가하기


# RMSE, R2 값 출력 (** 0.5 는 루트를 의미 --> 0.5 를 제곱으로 계산하겠다는 의미)
print('RMSE = ', )
print('R2 = ', )


---------------

### 알고리즘: 4) XGBoost

#### <XGBoost 장단점>

장점
 - 예측 속도가 상당히 빠르고, 예측력 또한 우수함
 - 변수 종류가 많고 데이터가 클수록 상대적으로 뛰어난 성능을 보임


단점
 - 복잡한 모델인만큼, 해석에 어려움이 있음
 - 더 나은 성능을 위한 하이퍼파라미터 튜닝이 까다로움

In [ ]:
from xgboost.sklearn import XGBRegressor

# 모델 불러오기
model_xgb =
# 모델 훈련하기


In [ ]:
# 학습된 모델을 사용하여 예측 및 평가하기
pred_xgb =

# RMSE, R2 값 출력 (** 0.5 는 루트를 의미 --> 0.5 를 제곱으로 계산하겠다는 의미)
print('RMSE = ', )
print('R2 = ', )

RMSE =  5313.776853763063
R2 =  0.8199313178273757
